src/languages/php.js

```diff
@@ -11,23 +11,105 @@ Category: common
  * @returns {LanguageDetail}
  * */
 export default function(hljs) {
+  const regex = hljs.regex;
+  const LABEL = '[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*' +
+    // negative look-ahead tries to avoid matching patterns that are not
+    // Perl at all like $ident$, @ident@, etc.
+    '(?![A-Za-z0-9])(?![$])';
+  const PASCAL_CASE_CLASS_NAME = regex.concat("[A-Z]", LABEL);
   const VARIABLE = {
-    className: 'variable',
-    begin: '\\$+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*' +
-      // negative look-ahead tries to avoid matching patterns that are not
-      // Perl at all like $ident$, @ident@, etc.
-      `(?![A-Za-z0-9])(?![$])`
+    scope: 'variable',
+    match: '\\$+' + LABEL,
+  };
+  const CONSTANT_CALL = regex.concat(LABEL, "\\b(?!\\()");
+  const CONSTANT = {
+    variants: [
+      {
+        match: [
+          /const/,
+          /\s+/,
+          LABEL,
+          /\s*=/,
+        ],
+        scope: {
+          1: "keyword",
+          3: "variable",
+        },
+      },
+      {
+        match: [
+          regex.concat(
+            /::/,
+            regex.lookahead(/(?!class\b)/)
+          ),
+          CONSTANT_CALL,
+        ],
+        scope: {
+          2: "variable",
+        },
+      },
+      {
+        match: [
+          PASCAL_CASE_CLASS_NAME,
+          regex.concat(
+            "::",
```

Comment:
regex please

Reply:
Can we just make a separate rule for `PASCAL_CASE_CLASS_NAME` entirely to highlight class names in ALL contexts?  That'd remove the need for a few of these, would it not?

Reply:
> regex please

You mean: change `"::"` to `/::/`?

> Can we just make a separate rule for PASCAL_CASE_CLASS_NAME entirely to highlight class names in ALL contexts? That'd remove the need for a few of these, would it not?

This could highlight code that is not a class. Like enum, constant, or keywords: Closure Generator. 
Doesn't sound good.

Reply:
> You mean: change "::" to /::/?

Yes.

> This could highlight code that is not a class. Like enum, constant, or keywords: Closure Generator.


I feel like i'd need examples in context to comment, pretend I'm not familiar with PHP. But lets take enum quickly (because I just googled it).

```
enum Status
{
    case DRAFT;
    case PUBLISHED;
    case ARCHIVED;
}
```

We could match `enum [klass]` with a special rule but also `Status` on it's own (by it's casing) would qualify as a `title.class` (all higher-level class-like things do)... the enum values are UPPER case so they  could be flagged as constants... now if someone wanted to use camel case for enum values then we would have a problem since they would be flagged as `title.class` vs `variable.constant`...

So perhaps this isn't a good idea, I don't know enough PHP to say for certain.

Of course all this talk of casing would require us making PHP case-sensitive and then dealing with the fall-out of that...